### PR TITLE
Separate kernel-dependent/tree-independent state out of wrangler

### PR DESCRIPTION
**Pull request by inducer**
Should go in at the same time as 
- [ ] https://github.com/inducer/sumpy/pull/70
- [ ] https://github.com/inducer/pytential/pull/101

cc @gaohao95

### CODE REVIEW

File: boxtree/fmm.py [hunk 1]
```diff
@@ -43,6 +43,33 @@
 
 # {{{ expansion wrangler interface
 
+# Design considerations:
+#
+# - Making the wrangler contain/permit it to depend the tree (which was previously the case)
+#   forces code caches (say, for translations) to be thrown away every time the FMM is run on
```
Comment (isuruf) [inline boxtree/fmm.py:49]: What do you mean by code caches here?

Comment (inducer) [reply boxtree/fmm.py:49]: Like sumpy's `wrangler.code.p2m`. Things that ultimately hold the reference to the finished loopy kernel (e.g.) doing the translation.

Comment (isuruf) [reply boxtree/fmm.py]: Got it. And `rotation_classes` would be cached here as well? (`translation_classes` depend on the tree, so those will have to be cached in taw)

**Comment by inducer** [comment]
Yup.

**Comment by inducer** [comment]
@gaohao95 @isuruf 0383c05 is a different, less invasive, less circular attempt at separating tree-independent from tree-dependent data. Take a look and let me know what you think.

File: boxtree/fmm.py [hunk 2]
```diff
@@ -20,234 +28,109 @@
...
 
-    local_result, timing_future = wrangler.eval_locals(
-            traversal.level_start_target_box_nrs,
-            traversal.target_boxes,
-            local_exps)
+        Wranglers should not hold a reference (and thereby be specific to) a
+        :class:`boxtree.Tree` instance. Their purpose is to host caches for
+        generated translation code that is reusable across trees.
+        It is OK for expansion wranglers to be specific to a given kernel
+        (or set of kernels).
```
Comment (gaohao95) [inline boxtree/fmm.py:66]: Does this comment still apply to the latest version? Looks like now `ExpansionWranglerInterface` holds both tree dependent and tree independent data, so should we move this comment to `TreeIndependentDataForWrangler`?

Comment (inducer) [reply boxtree/fmm.py]: Good catch! 7604d83

File: boxtree/fmm.py [hunk 3]
```diff
@@ -20,234 +28,116 @@
...
-
-    # {{{ "Stage 7:" propagate local_exps downward
-
-    local_exps, timing_future = wrangler.refine_locals(
-            traversal.level_start_target_or_target_parent_box_nrs,
-            traversal.target_or_target_parent_boxes,
-            local_exps)
-
-    recorder.add("refine_locals", timing_future)
+class ExpansionWranglerInterface(ABC):
```
[the remark this answers is not in the document]
Comment (inducer) [reply boxtree/fmm.py:61]: I'm a bit hesitant about that, but I could be convinced. What's the use case?

[the remark this answers is not in the document]
Comment (inducer) [reply boxtree/fmm.py:60]: I can see how the distributed wrangler would need to find where the multipoles live. Why not, let's make it public.

**Comment by inducer** [comment]
@gaohao95 FYI a43f5738c978f4ce8d6afffc0a87b925922324fd removes the `*zeros` methods from the wrangler interface, because they need to look different for sumpy: https://github.com/inducer/sumpy/pull/70/commits/d607136e6f56c18b68264449133dbe7ad0201af2. Will that cause a problem on your end?

**Comment by gaohao95** [comment]
> @gaohao95 FYI [a43f573](https://github.com/inducer/boxtree/commit/a43f5738c978f4ce8d6afffc0a87b925922324fd) removes the `*zeros` methods from the wrangler interface, because they need to look different for sumpy: [inducer/sumpy@d607136](https://github.com/inducer/sumpy/commit/d607136e6f56c18b68264449133dbe7ad0201af2). Will that cause a problem on your end?

I merged this upstream change in https://github.com/inducer/boxtree/pull/8/commits/8082857f2f51efcd3fb6ec7dc518a4841879db2d, and the test still works for me locally, so I think it should be okay, at least for distributed boxtree.

File: boxtree/fmm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (inducer) [reply boxtree/fmm.py:60]: c6b1d50

**Pull request merged**